### DISCUSSION TITLE
Custom middleware for unwrapping responses

### DISCUSSION THREAD

**Discussion by nmbrone** [elixir-tesla/tesla#474] [General]
Hi there 👋.

First of all many thanks for a great library.

I like to have a common response handler that does the response status check and unwraps the body.

It can be easily achieved by implementing a middleware:

```elixir
defmodule HandleResponse do
  @behaviour Tesla.Middleware

  @impl true
  def call(env, next, _opts) do
    env
    |> Tesla.run(next)
    |> case do
      {:ok, %Tesla.Env{status: status, body: body}} when status in 200..299 ->
        {:ok, body}

      {:ok, %Tesla.Env{body: body}} ->
        {:error, body}

      other ->
        other
    end
  end
end
```

Obviously, it breaks the middleware contract since it returns `{:ok, any()}` instead of `{:ok, %Tesla.Env{}}`. But as long as you keep such middleware at the tail of the response handlers chain it works perfectly fine and I'm okay with it.

The problem is that Dyalizer is not okay with it.

```elixir
defmodule Client do
  use Tesla

  plug HandleResponse
  # ... plug other middleware
end

{:ok, %{"json" => "here"}} = Client.get("url") # dialyzer gets crazy
```

Tesla already supports `docs: false` option. Maybe we could add `specs: false` option to tell
`Tesla.Builder` to not inject `@spec`s for `get`, `post` etc. functions?

Or maybe someone has a better idea of how to solve it?

**Comment by teamon** [comment]
ref https://github.com/teamon/tesla/issues/483
